sql, ui: collect, surface span stats for tables scanned under fingerprint plan

ericharmeling opened the pull request:
This commit adds an aggregated subset of the span stats for tables scanned during the execution of a fingerprint plan to the app stats collected for statement fingerprints.

To control how often we collect span stats, the commit also introduces a new SpanStatsCache.

Scanned span stats are surfaced in the Console as a new (hidden) column on the Statement Fingerprints Overview page that lists the "Percent Live Bytes" of all scanned tables and on the Statement Fingerprint Details page on the Overview tab. The stats are also recorded to the in-memory `crdb_internal.statement_statistics` table and eventually flushed to `system.statement_statistics`.

Console loom: https://www.loom.com/share/f4cca3a690134158aa4d84d9feaf9472

Fixes #72048.

Release note (ui change): The Statement Fingerprints Overview and Details pages now list the live bytes, total bytes, and the percent live/total bytes in tables scanned in the execution of a fingerprint's plan. The crdb_internal.statement_statistics table includes a new scannedSpanStats field with a subset of the MVCC stats for a statement plan's scanned tables.

Commit:
sql, ui: collect, surface span stats for tables scanned under fingerprint plan

This commit adds an aggregated subset of the span stats for tables scanned during
the execution of a fingerprint plan to the app stats collected for statement fingerprints.

To control how often we collect span stats, the commit also introduces a new SpanStatsCache.

Scanned span stats are surfaced in the Console as a new column on the Statement Fingerprints
Overview page that lists the "Percent Live Bytes" of all scanned tables and on the Statement
Fingerprint Details page on the Overview tab. The stats are also recorded to the in-memory
crdb_internal.statement_statistics table and eventually flushed to system.statement_statistics.

Fixes #72048.

Release note (ui change): The Statement Fingerprints Overview and Details pages now list the
live bytes, total bytes, and the percent live/total bytes in tables scanned in the execution
of a fingerprint's plan. The crdb_internal.statement_statistics table includes a new
scannedSpanStats field with a subset of the MVCC stats for a statement plan's scanned tables.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/98002)
<!-- Reviewable:end -->

ericharmeling commented:
Currently running `BenchmarkScannedSpanStats` at 5000x. Will post results here when that finishes.

yuzefovich reviewed (commented):
Nice! Glad to see more observability like this being exposed.

I did a quick drive-by read and got some comments / questions.

Also, are you aware of `./scripts/bench` for running benchmarks? That script runs some benchmark 10 times on two SHAs and compares the results. If you add another WIP commit which enables the scanned spans stats tracking (let's call that commit `HEAD`) and run `BENCHTIMEOUT=72h PKG=pkg/sql/sqlstats/sslocal/scannedspanstats BENCH=BenchmarkScannedSpanStats ./scripts/bench HEAD^ HEAD`, it'll print out the comparison of the old and the new commit.

Additionally, generally speaking it's a good idea to run benchmarks on the gceworker, just FYI.

> Reviewed 13 of 32 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-:-NPyo38w6xa5LMNm75Ns:bxcsttd)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ericharmeling)

___
*[`pkg/sql/conn_executor_exec.go` line 1410 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NPyjZ9S7YKy1WuK7e3I:-NPyjZ9S7YKy1WuK7e3J:bddwfl1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a0d8d291686f789ba3e6e3f89a0a0ece5045dc9/pkg/sql/conn_executor_exec.go#L1410)):*
> ```Go
> 	if planner.txn.IsOpen() {
> 		// Set index recommendations, so it can be saved on statement statistics.
> 		// TODO(yuzefovich): figure out whether we want to set isInternalPlanner
> ```

nit: move this TODO as well so that it stays right above the `isInternal := ...` line.
___
*[`pkg/sql/instrumentation.go` line 208 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NPyjkzC-oxn_PMVyh0J:-NPyjkzC-oxn_PMVyh0K:biops4j) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a0d8d291686f789ba3e6e3f89a0a0ece5045dc9/pkg/sql/instrumentation.go#L208)):*
> ```Go
> 	indexesUsed []string
> 
> 	// tablesScanned list the tables scanned by the execution of a plan.
> ```

nit: it'd be good to mention how a table is mapped to its `int64` (I'm assuming it's by TableID? but then why not use the proper TableID type?).
___
*[`pkg/sql/instrumentation.go` line 832 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NPyjxK25UORM0Jj2uD6:-NPyjxK25UORM0Jj2uD7:b8sr5a1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a0d8d291686f789ba3e6e3f89a0a0ece5045dc9/pkg/sql/instrumentation.go#L832)):*
> ```Go
> }
> 
> // CollectScannedSpanStats collects span stats for the tables scanned during the execution of the current statement's
> ```

nit: wrap comments at 80 characters.
___
*[`pkg/sql/instrumentation.go` line 844 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NPyk6FH1s9Y3fwpS0tE:-NPyk6FH1s9Y3fwpS0tF:bl8t6lf) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a0d8d291686f789ba3e6e3f89a0a0ece5045dc9/pkg/sql/instrumentation.go#L844)):*
> ```Go
> 		stmtType,
> 		isInternal,
> 		int64(ih.scanCounts[exec.ScanCount])) {
> ```

nit: it's a bit easier to read if the closing parenthesis of the function call is on a separate line, i.e. something like
```
	if spanStats.ShouldCacheFingerprintInfo(
		stmtType,
		isInternal,
		int64(ih.scanCounts[exec.ScanCount]),
	) {
```
___
*[`pkg/sql/instrumentation.go` line 864 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NPyl11q3N3EdR52bUVl:-NPyl11q3N3EdR52bUVm:b-7bjcz0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a0d8d291686f789ba3e6e3f89a0a0ece5045dc9/pkg/sql/instrumentation.go#L864)):*
> ```Go
> 					}
> 					spanStatsRequest := roachpb.SpanStatsRequest{
> 						NodeID:   strconv.Itoa(0),
> ```

nit: you can just use `"0"` if it's always zero. This will result in sending the `SpanStatsRequest` to all nodes in the cluster though, is this desirable?
___
*[`pkg/sql/appstatspb/app_stats.proto` line 132 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NPylTzoCJxC2iYYtYtx:-NPylTzoCJxC2iYYtYty:bqupppo) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a0d8d291686f789ba3e6e3f89a0a0ece5045dc9/pkg/sql/appstatspb/app_stats.proto#L132)):*
> ```protobuf
>   optional string last_error_code = 32 [(gogoproto.nullable) = false];
> 
>   // ScannedSpanStats is an aggregated subset of statistics for the spans of the tables scanned during the execution
> ```

nit: ditto for wrapping at 80 characters.
___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 899 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NPym2Fv-RE7U1ccf9n7:-NPym2Fv-RE7U1ccf9n8:b666ycr) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a0d8d291686f789ba3e6e3f89a0a0ece5045dc9/pkg/sql/opt/exec/execbuilder/relational.go#L899)):*
> ```Go
> 		return execPlan{}, err
> 	}
> 	b.TablesScanned = util.CombineUniqueInt64(b.TablesScanned, []int64{int64(tab.ID())})
> ```

Two places where we aggregate tables into `TablesScanned` right now are the "canonical scans", but there are other operators (e.g. `buildIndexJoin`) that also result in reads of data from disk. Are we ok with ignoring the latter?
___
*[`pkg/sql/sqlstats/sslocal/scannedspanstats/scanned_span_stats_cache.go` line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NPympef8wX_tRHVj9WU:-NPympegCtWmJmBBUXoM:bwtdno7) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a0d8d291686f789ba3e6e3f89a0a0ece5045dc9/pkg/sql/sqlstats/sslocal/scannedspanstats/scanned_span_stats_cache.go#L85)):*
> ```Go
> 
> // ShouldCollectSpanStatsForFingerprint returns true if the execution count for the fingerprint is above the minimum
> // and if no span stats have been collected for the fingerprint in the past an hour.
> ```

nit: `s/in the past an hour/in the past hour/`.
___
*[`pkg/sql/sqlstats/sslocal/scannedspanstats/scanned_span_stats_cache.go` line 197 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NPynKmGDXnK6pM_RCYQ:-NPynKmGDXnK6pM_RCYR:bev6p4q) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a0d8d291686f789ba3e6e3f89a0a0ece5045dc9/pkg/sql/sqlstats/sslocal/scannedspanstats/scanned_span_stats_cache.go#L197)):*
> ```Go
> 				spanStatsArray = append(spanStatsArray, statsInfo.spanStats)
> 			} else {
> 				return []enginepb.MVCCStats{}, false
> ```

The behavior here is a bit surprising - we're accumulating `spanStats` objects into an array, but if we don't find stats for one of the tables, we return nothing (rather than those that we did find). Consider adding some comments here.
___
*[`pkg/sql/sqlstats/sslocal/scannedspanstats/main_test.go` line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NPynyyf-8StyxgghJzO:-NPynyyf-8StyxgghJzP:btj0ctq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a0d8d291686f789ba3e6e3f89a0a0ece5045dc9/pkg/sql/sqlstats/sslocal/scannedspanstats/main_test.go#L1)):*
> ```Go
> // Copyright 2021 The Cockroach Authors.
> ```

nit: `s/2021/2023/`.


<!-- Sent from Reviewable.io -->

Commit:
server: extend span statistics endpoint

Extends: #96223

This PR extends the implementation of our SpanStats RPC endpoint to
fetch stats for multiple spans at once. By extending the endpoint, we
amortize the cost of the RPC's node fanout across all requested spans,
whereas previously, we were issuing a fanout per span requested.
Additionally, this change batches KV layer requests for ranges fully
contained by the span, instead of issuing a request per fully contained
range.

Release note: None

https://cockroachlabs.atlassian.net/browse/DOC-1355 #Informs:
https://github.com/cockroachdb/cockroach/issues/33316 #Epic: CRDB-8035

ericharmeling reviewed (commented):
TFTR!

I think I've gotten to all of your comments. For some, I had some follow-up questions. 

I was not aware of `./scripts/bench` though - thank you for pointing me to that! I'm going to try that approach and post back here with results.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-:-NQ77fgJ6RPzICTMViSU:b-im21ar)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`pkg/sql/conn_executor_exec.go` line 1410 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NPyjZ9S7YKy1WuK7e3I:-NQ6wexk22MZKB1rNQzh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a0d8d291686f789ba3e6e3f89a0a0ece5045dc9/pkg/sql/conn_executor_exec.go#L1410)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: move this TODO as well so that it stays right above the `isInternal := ...` line.
</blockquote></details>

Done.
___
*[`pkg/sql/instrumentation.go` line 208 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NPyjkzC-oxn_PMVyh0J:-NQ6ycIs5g_y0sHXvPVx:b-9fohz4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a0d8d291686f789ba3e6e3f89a0a0ece5045dc9/pkg/sql/instrumentation.go#L208)):*
> it'd be good to mention how a table is mapped to its int64 

Done.

> I'm assuming it's by TableID?

Correct.

> why not use the proper TableID type?

Initially I was using `[]uint64`, but was finding that type was incompatible with some existing functions (particularly, `CombineUniqueInt64`). Thinking about this though, it might make more sense to store this as a map of empty structs, which should make it more efficient to check for uniqueness. Then I'm putting the keys of that map back into an array for the cache logic. 

The object returned by `cat.Table.ID()` is of type `StableID`. Assuming that's the "proper TableID type", this would become `map[cat.StableID]struct{}`, and then `StableID`s all the way down from there. 

Does that sound reasonable?
___
*[`pkg/sql/instrumentation.go` line 832 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NPyjxK25UORM0Jj2uD6:-NQ7-LT_06NVmk7lY5yj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a0d8d291686f789ba3e6e3f89a0a0ece5045dc9/pkg/sql/instrumentation.go#L832)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: wrap comments at 80 characters.
</blockquote></details>

Done.
___
*[`pkg/sql/instrumentation.go` line 844 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NPyk6FH1s9Y3fwpS0tE:-NQ70HZ38OLrmQxeJcnO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a0d8d291686f789ba3e6e3f89a0a0ece5045dc9/pkg/sql/instrumentation.go#L844)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: it's a bit easier to read if the closing parenthesis of the function call is on a separate line, i.e. something like
```
	if spanStats.ShouldCacheFingerprintInfo(
		stmtType,
		isInternal,
		int64(ih.scanCounts[exec.ScanCount]),
	) {
```
</blockquote></details>

Done.
___
*[`pkg/sql/instrumentation.go` line 864 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NPyl11q3N3EdR52bUVl:-NQ70OOEBvhOPEakpiN7:b-l1t6xq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a0d8d291686f789ba3e6e3f89a0a0ece5045dc9/pkg/sql/instrumentation.go#L864)):*
> you can just use "0" if it's always zero. 

Done.

> This will result in sending the SpanStatsRequest to all nodes in the cluster though, is this desirable?

Correct. I originally was sending a request per node until we got a result, but I think we want to aggregate statistics across all replicas, right? IIUC, the `spanStatsFanOut` function in the span stats server does this https://github.com/cockroachdb/cockroach/blob/4fee2a42e5f2c10cf0bb5d422121243928a6f297/pkg/server/span_stats_server.go#L46 correctly.

FWIW, we'll be able to consolidate the `SpanStats` requests with this: https://github.com/cockroachdb/cockroach/pull/98490
___
*[`pkg/sql/appstatspb/app_stats.proto` line 132 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NPylTzoCJxC2iYYtYtx:-NQ7YgdT9Hx9h7CZNP5U:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a0d8d291686f789ba3e6e3f89a0a0ece5045dc9/pkg/sql/appstatspb/app_stats.proto#L132)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: ditto for wrapping at 80 characters.
</blockquote></details>

Done.
___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 899 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NPym2Fv-RE7U1ccf9n7:-NQ7SUce6_OtqB572DtJ:bjyzgd4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a0d8d291686f789ba3e6e3f89a0a0ece5045dc9/pkg/sql/opt/exec/execbuilder/relational.go#L899)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Two places where we aggregate tables into `TablesScanned` right now are the "canonical scans", but there are other operators (e.g. `buildIndexJoin`) that also result in reads of data from disk. Are we ok with ignoring the latter?
</blockquote></details>

Thanks for pointing this out! I think we do want all reads from disk (what I'm still thinking of as " span scans"). 

I've added the `TablesScanned` aggregation to the following:

- `buildIndexJoin`
- `buildLookupJoin`
- `buildInvertedJoin`
- `buildZigzagJoin`

Does this look right? Am I missing anything?
___
*[`pkg/sql/sqlstats/sslocal/scannedspanstats/scanned_span_stats_cache.go` line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NPympef8wX_tRHVj9WU:-NQ7Tx9H9KTwaylAZaYI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a0d8d291686f789ba3e6e3f89a0a0ece5045dc9/pkg/sql/sqlstats/sslocal/scannedspanstats/scanned_span_stats_cache.go#L85)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: `s/in the past an hour/in the past hour/`.
</blockquote></details>

Done.
___
*[`pkg/sql/sqlstats/sslocal/scannedspanstats/scanned_span_stats_cache.go` line 197 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NPynKmGDXnK6pM_RCYQ:-NQ7U2zh9CO04XjYyGiW:b-9gb76r) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a0d8d291686f789ba3e6e3f89a0a0ece5045dc9/pkg/sql/sqlstats/sslocal/scannedspanstats/scanned_span_stats_cache.go#L197)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

The behavior here is a bit surprising - we're accumulating `spanStats` objects into an array, but if we don't find stats for one of the tables, we return nothing (rather than those that we did find). Consider adding some comments here.
</blockquote></details>

My thinking here is that we don't want to give an incomplete picture of the MVCC stats from the tables scanned under a plan, as it could be misleading. 

I'm thinking of a scenario in which a very slow `SELECT` statement scans 2+ tables, but we are unable to get the stats for one of those tables, and perhaps that missing table has a lot of MVCC garbage that the storage iterator has scan over. If the table(s) for which we *were* able to get stats has very little MVCC garbage, then we'd report a much higher "live data percentage" in the statement statistics for that execution than we actually have, leading the user to believe that MVCC garbage is not a reason for the query's slowness, when it could have been.

Does that make sense? If so, I can add some comments. If not, maybe I'm thinking about this wrong.
___
*[`pkg/sql/sqlstats/sslocal/scannedspanstats/main_test.go` line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NPynyyf-8StyxgghJzO:-NQ74AoK-Lc-Pk3jlbai:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a0d8d291686f789ba3e6e3f89a0a0ece5045dc9/pkg/sql/sqlstats/sslocal/scannedspanstats/main_test.go#L1)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: `s/2021/2023/`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 10 of 14 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98002)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ericharmeling)

___
*[`pkg/sql/instrumentation.go` line 208 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NPyjkzC-oxn_PMVyh0J:-NQRZ2Iy7IG5ru6Ff4vL:bok37n3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a0d8d291686f789ba3e6e3f89a0a0ece5045dc9/pkg/sql/instrumentation.go#L208)):*
<details><summary><i>Previously, ericharmeling (Eric Harmeling) wrote…</i></summary><blockquote>

> it'd be good to mention how a table is mapped to its int64 

Done.

> I'm assuming it's by TableID?

Correct.

> why not use the proper TableID type?

Initially I was using `[]uint64`, but was finding that type was incompatible with some existing functions (particularly, `CombineUniqueInt64`). Thinking about this though, it might make more sense to store this as a map of empty structs, which should make it more efficient to check for uniqueness. Then I'm putting the keys of that map back into an array for the cache logic. 

The object returned by `cat.Table.ID()` is of type `StableID`. Assuming that's the "proper TableID type", this would become `map[cat.StableID]struct{}`, and then `StableID`s all the way down from there. 

Does that sound reasonable?
</blockquote></details>

Yeah, this looks good.
___
*[`pkg/sql/instrumentation.go` line 832 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NPyjxK25UORM0Jj2uD6:-NQRZO2i8O-FdEb1s-Fw:b-3j6ybj) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a0d8d291686f789ba3e6e3f89a0a0ece5045dc9/pkg/sql/instrumentation.go#L832)):*
<details><summary><i>Previously, ericharmeling (Eric Harmeling) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

BTW you could set up some helpers for this in GoLand - I have a visual guide at 80 characters but don't use the "hard wrap" when typing:

![Screenshot 2023-03-13 at 1.02.51 PM.png](https://files.reviewableusercontent.io/images/reviewable/5135385/5mqbLPozEGN6UymPhRC2vA58/Screenshot%202023-03-13%20at%201.02.51%20PM.png)
___
*[`pkg/sql/instrumentation.go` line 864 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NPyl11q3N3EdR52bUVl:-NQRZp3s7vdzWsHHH1HG:b-r933i8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a0d8d291686f789ba3e6e3f89a0a0ece5045dc9/pkg/sql/instrumentation.go#L864)):*
<details><summary><i>Previously, ericharmeling (Eric Harmeling) wrote…</i></summary><blockquote>

> you can just use "0" if it's always zero. 

Done.

> This will result in sending the SpanStatsRequest to all nodes in the cluster though, is this desirable?

Correct. I originally was sending a request per node until we got a result, but I think we want to aggregate statistics across all replicas, right? IIUC, the `spanStatsFanOut` function in the span stats server does this https://github.com/cockroachdb/cockroach/blob/4fee2a42e5f2c10cf0bb5d422121243928a6f297/pkg/server/span_stats_server.go#L46 correctly.

FWIW, we'll be able to consolidate the `SpanStats` requests with this: https://github.com/cockroachdb/cockroach/pull/98490
</blockquote></details>

My comment comes from the performance overhead angle - in large clusters (say 50 nodes) when this span stats collection is enabled, then we'd be issuing 50 RPCs for each statement for which we're collecting span stats. I haven't thought much about the problem we're solving (nor have I read this PR very carefully - let me know if you think I should be the primary reviewer), but this overhead might be prohibitively expensive if we do plan to enable the setting by default. I guess we should be issuing these RPCs only about once an hour, right? If so, then it seems reasonable and my concern is overblown.
___
*[`pkg/sql/instrumentation.go` line 848 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NQRZfZtC_9g9OoksJj9:-NQRZfZtC_9g9OoksJjA:b-z9yhl4) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdad5d29cca43c8739d8371ca05e08fcd7287063/pkg/sql/instrumentation.go#L848)):*
> ```Go
> 		var spanStatsArray []enginepb.MVCCStats
> 		newStatsForFingerprint := false
> 		var tablesScannedArray []cat.StableID
> ```

nit: you could precisely allocate this array given that we know `len(ih.tableScanned)`.
___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 899 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NPym2Fv-RE7U1ccf9n7:-NQR_nde5J8eimm5BWiU:bc1utib) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a0d8d291686f789ba3e6e3f89a0a0ece5045dc9/pkg/sql/opt/exec/execbuilder/relational.go#L899)):*
<details><summary><i>Previously, ericharmeling (Eric Harmeling) wrote…</i></summary><blockquote>

Thanks for pointing this out! I think we do want all reads from disk (what I'm still thinking of as " span scans"). 

I've added the `TablesScanned` aggregation to the following:

- `buildIndexJoin`
- `buildLookupJoin`
- `buildInvertedJoin`
- `buildZigzagJoin`

Does this look right? Am I missing anything?
</blockquote></details>

:thumbsup: I think that's all of the relevant places.
___
*[`pkg/sql/sqlstats/sslocal/scannedspanstats/scanned_span_stats_cache.go` line 197 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NPynKmGDXnK6pM_RCYQ:-NQRjrL_8Xf1EkuJmkTl:b-uiy35) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a0d8d291686f789ba3e6e3f89a0a0ece5045dc9/pkg/sql/sqlstats/sslocal/scannedspanstats/scanned_span_stats_cache.go#L197)):*
<details><summary><i>Previously, ericharmeling (Eric Harmeling) wrote…</i></summary><blockquote>

My thinking here is that we don't want to give an incomplete picture of the MVCC stats from the tables scanned under a plan, as it could be misleading. 

I'm thinking of a scenario in which a very slow `SELECT` statement scans 2+ tables, but we are unable to get the stats for one of those tables, and perhaps that missing table has a lot of MVCC garbage that the storage iterator has scan over. If the table(s) for which we *were* able to get stats has very little MVCC garbage, then we'd report a much higher "live data percentage" in the statement statistics for that execution than we actually have, leading the user to believe that MVCC garbage is not a reason for the query's slowness, when it could have been.

Does that make sense? If so, I can add some comments. If not, maybe I'm thinking about this wrong.
</blockquote></details>

This makes sense. Consider adding some of this explanation as a comment to the code.
___
*[`pkg/sql/sqlstats/sslocal/scannedspanstats/scanned_span_stats_cache.go` line 194 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NQRjOp03XwgV4j1fH0j:-NQRjOp03XwgV4j1fH0k:b-haaraj) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdad5d29cca43c8739d8371ca05e08fcd7287063/pkg/sql/sqlstats/sslocal/scannedspanstats/scanned_span_stats_cache.go#L194)):*
> ```Go
> 
> 	info, found := spanStats.getFingerprintInfoFromFingerprint(fingerprint)
> 	if found {
> ```

nit: perhaps it'd be a bit cleaner if we checked `!found` and return `nil, false` - then we'd unindent all the code below (which is the interesting part of this function).
___
*[`pkg/sql/sqlstats/sslocal/scannedspanstats/scanned_span_stats_cache.go` line 195 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NQRk2KcDbrOKiTF67nP:-NQRk2KcDbrOKiTF67nQ:b5gqhg7) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdad5d29cca43c8739d8371ca05e08fcd7287063/pkg/sql/sqlstats/sslocal/scannedspanstats/scanned_span_stats_cache.go#L195)):*
> ```Go
> 	info, found := spanStats.getFingerprintInfoFromFingerprint(fingerprint)
> 	if found {
> 		for _, tabID := range info.tableIDs {
> ```

nit: do you think that, generally speaking, we will find stats for all tables? If so, it'd be good to pre-allocate `spanStatsArray` precisely.


<!-- Sent from Reviewable.io -->

ericharmeling reviewed (commented):
I think I've responded to all of your comments. Thanks for the review!

I modified the bench test locally to not set the `scanned_span_stats_collection.enabled` cluster setting, and then ran the `bench` script as you suggested:

```
BENCHTIMEOUT=72h PKG=pkg/sql/sqlstats/sslocal/scannedspanstats BENCHES=BenchmarkScannedSpanStats ./scripts/bench HEAD^ HEAD

Comparing HEAD (new) with HEAD^ (old)
...
Switching to HEAD^
+ ./dev bench pkg/sql/sqlstats/sslocal/scannedspanstats --timeout=72h --filter=BenchmarkScannedSpanStats --count=10 --bench-mem -v --stream-output --ignore-cache
Switching to HEAD
+ ./dev bench pkg/sql/sqlstats/sslocal/scannedspanstats --timeout=72h --filter=BenchmarkScannedSpanStats --count=10 --bench-mem -v --stream-output --ignore-cache
...
```

Here are the aggregated mean results:

||bench.HEAD^ | bench.HEAD | % vs base|
| --- | --- | --- | --- |
|sec/op  | 47.28m  | 50.05m | +5.86%
|B/op  | 20.37Mi | 22.77Mi  | +11.75%
|allocs/op | 116.8k | 127.0k  | +8.73%

It's worth noting that we limit the number of tables that have span stats in the cache to 100 by default. One of the benchmark configs moves this up to 500, and those results really push up the average in latency and memory. So these results are skewed a bit to the larger/slower side.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-:-NQvx0Ec6KgEXAW2BvQp:blla7dn)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`pkg/sql/instrumentation.go` line 832 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NPyjxK25UORM0Jj2uD6:-NQlPtxi8sW2qtpQjK3I:bu2mcne) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a0d8d291686f789ba3e6e3f89a0a0ece5045dc9/pkg/sql/instrumentation.go#L832)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

BTW you could set up some helpers for this in GoLand - I have a visual guide at 80 characters but don't use the "hard wrap" when typing:

![Screenshot 2023-03-13 at 1.02.51 PM.png](https://files.reviewableusercontent.io/images/reviewable/5135385/5mqbLPozEGN6UymPhRC2vA58/Screenshot%202023-03-13%20at%201.02.51%20PM.png)

</blockquote></details>

Thanks for the tip! Done.
___
*[`pkg/sql/instrumentation.go` line 864 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NPyl11q3N3EdR52bUVl:-NQw93b29mnMVOWcCZV-:bkayv3l) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a0d8d291686f789ba3e6e3f89a0a0ece5045dc9/pkg/sql/instrumentation.go#L864)):*
> in large clusters (say 50 nodes) when this span stats collection is enabled, then we'd be issuing 50 RPCs for each statement for which we're collecting span stats. 

We'd actually be issuing more RPCs than that, since for each statement, we are sending an RPC *per table*. So in the worst-case for 50 nodes, that would be 50\*nStatement\*nTables. 

>this overhead might be prohibitively expensive if we do plan to enable the setting by default.

That's a really valid concern (in fact, my main concern with a PR like this). With that in mind, the plan is to disable span stats collection by default (as seen in this PR). When manually enabled, we only issue the RPCs every hour (as you've already noted) which should significantly limit the impact on performance.

I think https://github.com/cockroachdb/cockroach/pull/98490 will make these span stats calls much more performant, especially on large clusters, as we'll be able to limit the fanout to one per statement, rather than per table per statement. We could look into shortening that interval if after further testing we find the extended span stats server to improve performance in a meaningful way.

> nor have I read this PR very carefully - let me know if you think I should be the primary reviewer

I'd definitely appreciate a closer look if you have the bandwidth, but someone on Cluster Obs will also need to review this.
___
*[`pkg/sql/instrumentation.go` line 848 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NQRZfZtC_9g9OoksJj9:-NQw8s455wd4SP6zYyP3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdad5d29cca43c8739d8371ca05e08fcd7287063/pkg/sql/instrumentation.go#L848)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: you could precisely allocate this array given that we know `len(ih.tableScanned)`.
</blockquote></details>

Done.
___
*[`pkg/sql/sqlstats/sslocal/scannedspanstats/scanned_span_stats_cache.go` line 197 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NPynKmGDXnK6pM_RCYQ:-NQwIDBT75T2kDuzijcr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a0d8d291686f789ba3e6e3f89a0a0ece5045dc9/pkg/sql/sqlstats/sslocal/scannedspanstats/scanned_span_stats_cache.go#L197)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

This makes sense. Consider adding some of this explanation as a comment to the code.
</blockquote></details>

Done.
___
*[`pkg/sql/sqlstats/sslocal/scannedspanstats/scanned_span_stats_cache.go` line 194 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NQRjOp03XwgV4j1fH0j:-NQwG185-Fn2-4Ep4NUa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdad5d29cca43c8739d8371ca05e08fcd7287063/pkg/sql/sqlstats/sslocal/scannedspanstats/scanned_span_stats_cache.go#L194)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: perhaps it'd be a bit cleaner if we checked `!found` and return `nil, false` - then we'd unindent all the code below (which is the interesting part of this function).
</blockquote></details>

Done.
___
*[`pkg/sql/sqlstats/sslocal/scannedspanstats/scanned_span_stats_cache.go` line 195 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NQRk2KcDbrOKiTF67nP:-NQwG1kF4tMgEPXkw4qB:b5cn2pf) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdad5d29cca43c8739d8371ca05e08fcd7287063/pkg/sql/sqlstats/sslocal/scannedspanstats/scanned_span_stats_cache.go#L195)):*
> do you think that, generally speaking, we will find stats for all tables?

I think we will. But there could be a case of some table stats being empty because we were unable to get their span stats from the span stats server.

> it'd be good to pre-allocate spanStatsArray

Done.


<!-- Sent from Reviewable.io -->

THardy98 reviewed (commented):
Haven't looked at the code too closely but just some thoughts on the design after some discussion offline.

Effectively we have a cache of span statistics for tables (by ID), and an indexing structure to aggregate them by statement fingerprint.

We update the cache at each statement we encounter. We check for stale stats (>1 hour or whatever the cluster setting is set to), and if there is room in the cache. If both conditions pass, we fetch span statistics and update the table -> span stats cache with fresh span stats.

I'm thinking we could probably simplify the cache update by instead of updating at each statement encounter, we update at each "expiry interval" (i.e. the hour mark). We could collect all the unique table IDs that we encounter from all the statements that occur within the hour. With each table ID we encounter, all we do is update our indexing structure (fingerprint -> []tableIDs). At the hour interval, we can fetch the span statistics for all the unique table IDs we have encountered and update our cache (tableID -> span stats). This does a few things:
- We ensure we have complete MVCC stats for all fingerprints. If we want to enforce a table limit (primarily for memory/latency constraints), we can stop at the *fingerprint-level*. For example, we encounter `fingerprintB` that has 7 new table IDs that are not part of the cache. We check the cache and it only has room for 3 new table IDs, so we do not add *any* of them. Every fingerprint *prior* to `fingerprintB` will have span statistics cached for all their tables, and we avoid fetching incomplete MVCC stats on `fingerprintB`.
- We can batch easily (plays into this PR: https://github.com/cockroachdb/cockroach/pull/98490 nicely)
- Simplifies the logic. We know exactly which table IDs need fetching, and that all existing span statistics are fresh for the hour.
- Single point update, update time is consistent across all fingerprints.

I think there are 2 main drawbacks to this approach:
- A fingerprint's span statistics aren't fetched immediately, we have to wait till the update interval. Not that big of a drawback imo, as the UI surfaces SQL activity at the hour interval anyways.
- Single point update, we put the load of issuing all these fanouts at a single point in time. This would be rectified by: https://github.com/cockroachdb/cockroach/pull/98490 (maybe we could leave it off until it gets in?)

WDYT?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-:-NQzrbhh8bdRa8BSZyJp:b-cts5qc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 7 of 12 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98002)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ericharmeling)

___
*[`pkg/sql/instrumentation.go` line 844 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NR-l2sX4qZlJ2VeijVU:-NR-l2sX4qZlJ2VeijVV:blj5l5i) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd2465d106b4d71aaeeb4f469954308bafd0a7c4/pkg/sql/instrumentation.go#L844)):*
> ```Go
> 		stmtType,
> 		isInternal,
> 		int64(ih.scanCounts[exec.ScanCount]),
> ```

`ih.scanCounts` is specifically about the number of "canonical scans" (i.e. ignoring the index joins, lookup joins, etc) - is this what we want here? I believe so but thought I'd ask.
___
*[`pkg/sql/instrumentation.go` line 860 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NR-fYtj25Qpu5kUQbus:-NR-fYtj25Qpu5kUQbut:b-5pr9o5) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd2465d106b4d71aaeeb4f469954308bafd0a7c4/pkg/sql/instrumentation.go#L860)):*
> ```Go
> 					resolvedSpanStartKey, err := keys.Addr(tableStartKey)
> 					if err != nil {
> 						log.Warningf(ctx, "failed to resolve addr for key %q: %+v", tableStartKey, err)
> ```

If the resolution failed, shouldn't we short-circuit this method or skip this iteration of the loop?

Also, nit: perhaps add more context into the error message, like `for key %q when collecting span stats: %+v`.
___
*[`pkg/sql/sqlstats/sslocal/scannedspanstats/scanned_span_stats_cache.go` line 37 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NR-j0GE49Or_c3G25Ob:-NR-j0GE49Or_c3G25Oc:bdg8xrk) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd2465d106b4d71aaeeb4f469954308bafd0a7c4/pkg/sql/sqlstats/sslocal/scannedspanstats/scanned_span_stats_cache.go#L37)):*
> ```Go
> }
> 
> // SpanStatsCache stores the map of fingerprints to table IDs and the map of table IDs to span stats and execution count.
> ```

nit: there are some long comment lines in this file. (BTW I don't remember if I mentioned this before, but my nits about the comment line lengths come from our [style guide](https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/181371303/Go+Golang+coding+guidelines#Line-Length) - I don't want to sound like someone who is nit-picking for no reason :smile: (Granted, I'm probably more religious about this than most other people.))
___
*[`pkg/sql/sqlstats/sslocal/scannedspanstats/scanned_span_stats_cache.go` line 48 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NR-jcv6CbZKEXxlkHuR:-NR-jcv6CbZKEXxlkHuS:bxelaej) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd2465d106b4d71aaeeb4f469954308bafd0a7c4/pkg/sql/sqlstats/sslocal/scannedspanstats/scanned_span_stats_cache.go#L48)):*
> ```Go
> 		syncutil.RWMutex
> 
> 		// statsMap maps table IDs to span statistics
> ```

nit: missing period here and below.
___
*[`pkg/sql/sqlstats/sslocal/scannedspanstats/scanned_span_stats_cache.go` line 72 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NR-jsDE24ZX_FAFHgii:-NR-jsDE24ZX_FAFHgij:bbzmfwe) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd2465d106b4d71aaeeb4f469954308bafd0a7c4/pkg/sql/sqlstats/sslocal/scannedspanstats/scanned_span_stats_cache.go#L72)):*
> ```Go
> 	spanStatsCache.mu.statsMap = make(map[cat.StableID]spanStatsInfo)
> 	spanStatsCache.mu.fingerprintMap = make(map[string]fingerprintInfo)
> 	spanStatsCache.mu.lastCleanupTs = time.Time{}
> ```

nit: this is unnecessary - the default value of `time.Time` is exactly the same.
___
*[`pkg/sql/sqlstats/sslocal/scannedspanstats/scanned_span_stats_cache.go` line 87 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-NR-mamW1WKvHVSfLjDU:-NR-mamW1WKvHVSfLjDV:bx74vqb) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd2465d106b4d71aaeeb4f469954308bafd0a7c4/pkg/sql/sqlstats/sslocal/scannedspanstats/scanned_span_stats_cache.go#L87)):*
> ```Go
> // ShouldCollectSpanStatsForFingerprint returns true if the execution count for the fingerprint is above the minimum
> // and if no span stats have been collected for the fingerprint in the past hour.
> // If no cache exists for the fingerprint, it creates one and returns true.
> ```

nit: perhaps "no cached info" instead of "no cache" here and below?


<!-- Sent from Reviewable.io -->

ericharmeling reviewed (commented):
@THardy98
 
Thanks for the thoughtful design review! 

I really like your suggested approach. I think it makes a lot of sense, provided https://github.com/cockroachdb/cockroach/pull/98490 is merged. The performance of this feature currently benefits from RPCs being spread out. Without 98490, the spike from this many fan-outs at the exact same time will not be good. 

You mention this in your "2 main drawbacks". Re: the second one:

> A fingerprint's span statistics aren't fetched immediately, we have to wait till the update interval. Not that big of a drawback imo, as the UI surfaces SQL activity at the hour interval anyways.

I actually think this is a good thing, because we won't be underreporting MVCC garbage so drastically in the first hour. Collecting at a very low execution count could leave the picture pretty off-base after 45-55 minutes, especially if we're looking at a new workload against a new schema. But maybe that's not a super common situation.

I'm going to probably going to cut another branch, implement some changes in line with your suggestion (in addition to some of Yahor's latest comments), rebase on 98490, and then report the benchmarks.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98002#-:-NR--X0d4p-EGeTjG2Vj:b-l1jgxb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

Commit:
revision
